src/Compiler/Facilities/DiagnosticsLogger.fs

```diff
@@ -376,27 +376,20 @@ type CapturingDiagnosticsLogger(nm, ?eagerFormat) =
 
 /// Type holds thread-static globals for use by the compiler.
 type internal DiagnosticsThreadStatics =
-    [<ThreadStatic; DefaultValue>]
-    static val mutable private buildPhase: BuildPhase
+    static let buildPhase = new ThreadLocal<BuildPhase>(fun () -> BuildPhase.DefaultPhase)
 
-    [<ThreadStatic; DefaultValue>]
-    static val mutable private diagnosticsLogger: DiagnosticsLogger
+    static let diagnosticsLogger =
+        new ThreadLocal<DiagnosticsLogger>(fun () -> AssertFalseDiagnosticsLogger)
```

Comment:
Do we want it per thread or per async context/task? Because if it's latter, if we are getting executed in "task" (like we do in VS), ThreadLocal won't work, and we will need the AsyncLocal instead.